pkg/kubelet/cm/dra/manager.go

```diff
@@ -105,47 +106,62 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 						pod.Name, pod.UID, podResourceClaim.Name, resourceClaim.UID)
 				}
 
-				// Call NodePrepareResource RPC
-				driverName := resourceClaim.Status.DriverName
-
-				client, err := dra.NewDRAPluginClient(driverName)
-				if err != nil {
-					return fmt.Errorf("failed to get DRA Plugin client for plugin name %s, err=%+v", driverName, err)
+				resourceHandles := resourceClaim.Status.Allocation.Context
+				if len(resourceHandles) == 0 {
+					resourceHandles = []resourcev1alpha2.ResourceHandle{{KubeletPluginName: resourceClaim.Status.DriverName}}
 				}
 
-				response, err := client.NodePrepareResource(
-					context.Background(),
-					resourceClaim.Namespace,
-					resourceClaim.UID,
-					resourceClaim.Name,
-					resourceClaim.Status.Allocation.ResourceHandle)
-				if err != nil {
-					return fmt.Errorf("NodePrepareResource failed, claim UID: %s, claim name: %s, resource handle: %s, err: %+v",
-						resourceClaim.UID, resourceClaim.Name, resourceClaim.Status.Allocation.ResourceHandle, err)
-				}
-
-				klog.V(3).InfoS("NodePrepareResource succeeded", "response", response)
-
-				// NOTE: Passing CDI device names as annotations is a temporary solution
-				// It will be removed after all runtimes are updated
-				// to get CDI device names from the ContainerConfig.CDIDevices field
-				annotations, err := generateCDIAnnotations(resourceClaim.UID, driverName, response.CdiDevices)
-				if err != nil {
-					return fmt.Errorf("failed to generate container annotations, err: %+v", err)
+				var cdiDevices []string
+				var cdiAnnotations []kubecontainer.Annotation
+				for _, resourceHandle := range resourceHandles {
```

Comment:
I am going to need to do something similar on the `unprepare()` side. Which means I will need to cache the `cdiDevices` per `kubeletplugin` so that I can disambiguate them later.

Reply:
Done.